File: backend/app/models/knowledge.py
from datetime import datetime

from sqlalchemy import DateTime, Integer, String, Text
from sqlalchemy.dialects.mysql import LONGTEXT
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class KnowledgeDocument(Base):
    __tablename__ = "knowledge_documents"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    title: Mapped[str] = mapped_column(String(255), index=True)
    category: Mapped[str] = mapped_column(String(120), index=True)
    target_position: Mapped[str] = mapped_column(String(160), index=True)
    content: Mapped[str] = mapped_column(LONGTEXT().with_variant(Text, "sqlite"))
    metadata_json: Mapped[str] = mapped_column(Text)
    original_oss_key: Mapped[str | None] = mapped_column(String(512), nullable=True)
    parsed_text_oss_key: Mapped[str | None] = mapped_column(String(512), nullable=True)
    file_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    file_type: Mapped[str | None] = mapped_column(String(32), nullable=True)
    file_size: Mapped[int | None] = mapped_column(Integer, nullable=True)
    chunk_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    index_version: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    parse_status: Mapped[str] = mapped_column(String(32), default="completed", nullable=False)
    parse_error: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


class KnowledgeReindexJob(Base):
    __tablename__ = "knowledge_reindex_jobs"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    status: Mapped[str] = mapped_column(String(32), default="pending", nullable=False, index=True)
    source_collection: Mapped[str | None] = mapped_column(String(255), nullable=True)
    target_collection: Mapped[str | None] = mapped_column(String(255), nullable=True)
    total_documents: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    processed_documents: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    total_chunks: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    error: Mapped[str | None] = mapped_column(Text, nullable=True)
    recovery_strategy: Mapped[str] = mapped_column(
        Text,
        default="Keep the active alias on the previous collection; retry with a new collection.",
        nullable=False,
    )
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    started_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)

class KnowledgeIngestionTask(Base):
    __tablename__ = "knowledge_ingestion_tasks"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    document_id: Mapped[int | None] = mapped_column(Integer, nullable=True, index=True)
    status: Mapped[str] = mapped_column(String(32), default="pending", nullable=False, index=True)
    stage: Mapped[str] = mapped_column(String(32), default="queued", nullable=False)
    source_oss_key: Mapped[str] = mapped_column(String(512), nullable=False)
    parsed_text_oss_key: Mapped[str | None] = mapped_column(String(512), nullable=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    category: Mapped[str] = mapped_column(String(120), nullable=False)
    target_position: Mapped[str] = mapped_column(String(160), nullable=False)
    file_name: Mapped[str] = mapped_column(String(255), nullable=False)
    file_type: Mapped[str] = mapped_column(String(32), nullable=False)
    file_size: Mapped[int] = mapped_column(Integer, nullable=False)
    attempts: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    max_attempts: Mapped[int] = mapped_column(Integer, default=3, nullable=False)
    error: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    started_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)

File: frontend/src/api/knowledge.ts
import { apiClient } from './client'

export interface KnowledgeDocument {
  id: number
  title: string
  category: string
  target_position: string
  content: string
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface KnowledgeDocumentPayload {
  title: string
  category: string
  target_position: string
  content: string
  metadata: Record<string, unknown>
}

export interface KnowledgeIngestionTask {
  id: number
  document_id: number | null
  status: string
  stage: string
  title: string
  category: string
  target_position: string
  file_name: string
  file_type: string
  file_size: number
  attempts: number
  max_attempts: number
  error: string | null
  created_at: string
  updated_at: string
  started_at: string | null
  finished_at: string | null
}

export function fetchKnowledgeDocuments() { return apiClient.get<KnowledgeDocument[]>('/knowledge/documents') }
export function createKnowledgeDocument(data: KnowledgeDocumentPayload) { return apiClient.post<KnowledgeDocument>('/knowledge/documents', data) }
export function updateKnowledgeDocument(id: number, data: KnowledgeDocumentPayload) { return apiClient.put<KnowledgeDocument>(`/knowledge/documents/${id}`, data) }
export function deleteKnowledgeDocument(id: number) { return apiClient.delete(`/knowledge/documents/${id}`) }

export function uploadKnowledgeFile(data: { file: File; title?: string; category: string; target_position?: string }) {
  const form = new FormData()
  form.append('file', data.file)
  if (data.title?.trim()) form.append('title', data.title.trim())
  form.append('category', data.category)
  if (data.target_position?.trim()) form.append('target_position', data.target_position.trim())
  return apiClient.post<KnowledgeIngestionTask>('/knowledge/files/async', form)
}

export function fetchIngestionTasks() { return apiClient.get<KnowledgeIngestionTask[]>('/knowledge/ingestion-tasks') }
export function fetchIngestionTask(id: number) { return apiClient.get<KnowledgeIngestionTask>(`/knowledge/ingestion-tasks/${id}`) }
export function retryIngestionTask(id: number) { return apiClient.post<KnowledgeIngestionTask>(`/knowledge/ingestion-tasks/${id}/retry`) }
export function reexecuteIngestionTask(id: number) { return apiClient.post<KnowledgeIngestionTask>(`/knowledge/ingestion-tasks/${id}/reexecute`) }
